Fix_reference_sorting

alvazir opened the pull request:
Add moved_cell key(reversed) with highest priority, because MVRF records should come first.

These few lines took me 3 hard days of debugging :-)

Greatness7 commented:
Hi, thanks for the PR! Could you provide a (minimal) example file that shows the loading problems? Just for science. :)

alvazir commented:
The problem is it's hard to notice it at first glance :-)

The minimal example would be(I may provide more complicated test cases if needed):
 1. Take "Balmora Waterworks.esp"(that's how I noticed it). It contains a lot of MVRFs.
 2. Then resave it with load_path and save_path. 
 3. Then go into a game(tested with OpenMW 0.47) with new plugin loaded.
 
 "coe -3 -2". You will notice missing instances around Odai river, misplaced doors and windows etc. Nothing in log though.
 
 Cell "Balmora (-3, -2)" would show changed instances order. MVRFs in a new plugin are scattered all around, unlike original plugin, where MVRFs are stacked at top(beginning) of references list. Resaving new plugin in TES3-CS would place those MVRFs at top again, and everything works again.
 
 UESP also says that MVRFs should come [first](https://en.m.uesp.net/wiki/Morrowind_Mod:Mod_File_Format/CELL). That's what I experience in CS behaviour.

Greatness7 commented:
> otherwise reference processing in both engines breaks

I was assuming this implied that loading breaks in both engine? I believe Morrowind/TESCS do not care about the order of MVRF. If they did then loading/saving in TESCS would not fix it, since you would've had incorrect data after the load. I did test this on OpenMW awhile ago and found it to be broken, and reported to the developers, but I guess it never got fixed and slipped my mind by the time reference sorting was implemented here.

Even if it's technically only an OpenMW bug, making the library output compatible files is still a good thing of course. 👍

Greatness7 commented:
Actually the bug did get fixed in OpenMW. A year ago. Just has not released yet.

https://gitlab.com/OpenMW/openmw/-/commit/3588bfcb174ecca5c8499f4c92d65571ff8270e4

(Merging your fix is still a good idea, since many users will be on older version of OpenMW.)

alvazir commented:
I am sorry for misinformation about MVRF records order breaking Morrowind.exe. That was a too broad assumption on my side :-\ I was convinced judging by OpenMW behaviour, TES3-CS output and UESP page...

I see two positive things in implementing MVRF priority sorting:
1. 0.47(and older) versions of OpenMW would be able to use results of something touching moved references with the library.
2. Less differences between files produced with the library and TES3-CS. The closer tes3cmd output of those the better.

I am interested in (1.) mostly, because I'm making an utility for merging plugins. MVRF issue renders utility unusable for 0.47(and older) :-\ Perhaps the change may be done via features?

There is also a slightly negative side(is it?). Perhaps performance would drop a tiny bit, though it should be neglible.

Greatness7 commented:
Yes I agree, supporting 0.47 is ideal, so I do support this change. If the sorting logic is going to be modified I would prefer we try to match the TESCS as much as possible. If I remember correctly it places MVRF's directly before the reference they affect, rather than all at the front of this list. I have not confirmed this yet though.

alvazir commented:
EDIT: slightly reworded, changed **tes3cmd dump** to **tes3cmd modify --type cell --exact-id "balmora (-3, -2)" --run "\$R->dump()"** for better visibility.

Yes, it places MVRF right directly before the reference they affect. That's what library already does too with or without proposed changes.

Sorting change only moves those MVRF + FRMR pairs at the beginning. Example tes3cmd dumps for mentioned Balmora Waterworks:

1. Original(TES-CS produced, 4 pairs of MVRF + FRMR):

```
Record: CELL "balmora (-3, -2)" Flags:0x0000 ()
 [0] NAME: Name:Balmora
 [1] DATA: (Exterior) Coordinates: (-3, -2)  Flags:0x0006 (Has_Water, Illegal_To_Sleep_Here)
 [2] RGNN: Region:West Gash Region
 [3] MVRF: ObjIdx:7909  MastIdx:1
 [4] CNDT: X:-3  Y:-3 (External Cell Destination)
 [5] FRMR: ObjIdx:7909  MastIdx:1
 [6] NAME: Name:ex_hlaalu_steps_03
 [7] DATA: X:-20557.490  Y:-17064.256  Z:326.408  X_Angle:0.0000  Y_Angle:0.0000  Z_Angle:0.0000
 [8] MVRF: ObjIdx:41478  MastIdx:1
 [9] CNDT: X:-3  Y:-3 (External Cell Destination)
 [10] FRMR: ObjIdx:41478  MastIdx:1
 [11] NAME: Name:furn_de_lightpost_01
 [12] DELE: Deleted:4729956
 [13] MVRF: ObjIdx:19085  MastIdx:1
 [14] CNDT: X:-3  Y:-3 (External Cell Destination)
 [15] FRMR: ObjIdx:19085  MastIdx:1
 [16] NAME: Name:ex_hlaalu_win_01
 [17] DATA: X:-18162.238  Y:-16514.043  Z:476.361  X_Angle:0.0000  Y_Angle:0.0000  Z_Angle:1.5000
 [18] MVRF: ObjIdx:41164  MastIdx:1
 [19] CNDT: X:-3  Y:-3 (External Cell Destination)
 [20] FRMR: ObjIdx:41164  MastIdx:1
 [21] NAME: Name:ex_hlaalu_dsteps_03
 [22] DATA: X:-18218.293  Y:-16398.596  Z:337.681  X_Angle:0.0000  Y_Angle:0.0000  Z_Angle:1.5000
 [23] MVRF: ObjIdx:41617  MastIdx:1
 [24] CNDT: X:-3  Y:-3 (External Cell Destination)
 [25] FRMR: ObjIdx:41617  MastIdx:1
 [26] NAME: Name:hlaalu_loaddoor_02_MwG_DGB
 [27] DODT: X:-1354.89  Y:845.76  Z:-669.08  X_Angle:0.00  Y_Angle:0.00  Z_Angle:4.71
 [28] DNAM: Destination:Balmora, Dura gra-Bol's House
 [29] DATA: X:-18126.828  Y:-16392.756  Z:442.662  X_Angle:0.0000  Y_Angle:0.0000  Z_Angle:1.4992
 [30] FRMR: ObjIdx:41610  MastIdx:1
  ... other non-moved records
```

2. Resaved with proposed changes(same result, only 41478 MVRF + FRMR pair comes in slightly different order):

```
Record: CELL "balmora (-3, -2)" Flags:0x0000 ()
 [0] NAME: Name:Balmora
 [1] DATA: (Exterior) Coordinates: (-3, -2)  Flags:0x0006 (Has_Water, Illegal_To_Sleep_Here)
 [2] RGNN: Region:West Gash Region
 [3] MVRF: ObjIdx:7909  MastIdx:1
 [4] CNDT: X:-3  Y:-3 (External Cell Destination)
 [5] FRMR: ObjIdx:7909  MastIdx:1
 [6] NAME: Name:ex_hlaalu_steps_03
 [7] DATA: X:-20557.490  Y:-17064.256  Z:326.408  X_Angle:0.0000  Y_Angle:0.0000  Z_Angle:0.0000
 [8] MVRF: ObjIdx:19085  MastIdx:1
 [9] CNDT: X:-3  Y:-3 (External Cell Destination)
 [10] FRMR: ObjIdx:19085  MastIdx:1
 [11] NAME: Name:ex_hlaalu_win_01
 [12] DATA: X:-18162.238  Y:-16514.043  Z:476.361  X_Angle:0.0000  Y_Angle:0.0000  Z_Angle:1.5000
 [13] MVRF: ObjIdx:41164  MastIdx:1
 [14] CNDT: X:-3  Y:-3 (External Cell Destination)
 [15] FRMR: ObjIdx:41164  MastIdx:1
 [16] NAME: Name:ex_hlaalu_dsteps_03
 [17] DATA: X:-18218.293  Y:-16398.596  Z:337.681  X_Angle:0.0000  Y_Angle:0.0000  Z_Angle:1.5000
 [18] MVRF: ObjIdx:41478  MastIdx:1
 [19] CNDT: X:-3  Y:-3 (External Cell Destination)
 [20] FRMR: ObjIdx:41478  MastIdx:1
 [21] NAME: Name:furn_de_lightpost_01
 [22] DELE: Deleted:0
 [23] MVRF: ObjIdx:41617  MastIdx:1
 [24] CNDT: X:-3  Y:-3 (External Cell Destination)
 [25] FRMR: ObjIdx:41617  MastIdx:1
 [26] NAME: Name:hlaalu_loaddoor_02_MwG_DGB
 [27] DODT: X:-1354.89  Y:845.76  Z:-669.08  X_Angle:0.00  Y_Angle:0.00  Z_Angle:4.71
 [28] DNAM: Destination:Balmora, Dura gra-Bol's House
 [29] DATA: X:-18126.828  Y:-16392.756  Z:442.662  X_Angle:0.0000  Y_Angle:0.0000  Z_Angle:1.4992
 [30] FRMR: ObjIdx:7936  MastIdx:1
   ... other non-moved records
```

3. Resaved without proposed changes(MVRF + FRMR pairs are scattered all around, tes3cmd dump of this is going to be too large).
